From: Jiri Slaby <jslaby@suse.cz>
Date: Mon, 12 Feb 2018 16:09:25 +0100
Subject: ext4: llseek, do not crop offset on 32bit
Patch-mainline: Submitted on 12/02/2018
References: bsc#1079747

On 32bit, when llseek(SEEK_HOLE/DATA) passes down the start larger
than 32bit (like 0x100400000 in the tar testsuite), it overflows in
ext4_iomap_begin and gets cropped (to 0x400000). This indeed makes the
tar testsuite to fail.

So fix it by using the original offset.

Fixes: 545052e9e35a ("ext4: Switch to iomap for SEEK_HOLE / SEEK_DATA")
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 fs/ext4/inode.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -3523,7 +3523,7 @@ retry:
 		iomap->flags |= IOMAP_F_DIRTY;
 	iomap->bdev = inode->i_sb->s_bdev;
 	iomap->dax_dev = sbi->s_daxdev;
-	iomap->offset = first_block << blkbits;
+	iomap->offset = offset;
 	iomap->length = (u64)map.m_len << blkbits;
 
 	if (ret == 0) {
